server/src/main/java/org/opensearch/search/DefaultSearchContext.java

```diff
@@ -876,11 +877,23 @@ public Profilers getProfilers() {
     }
 
     /**
-     * Returns concurrent segment search status for the search context
+     * Returns concurrent segment search status for the search context. This should only be used after request parsing, during which requestShouldUseConcurrentSearch will be set.
      */
     @Override
-    public boolean isConcurrentSegmentSearchEnabled() {
-        return useConcurrentSearch;
+    public boolean shouldUseConcurrentSearch() {
+        assert requestShouldUseConcurrentSearch != null : "requestShouldUseConcurrentSearch must be set";
+        return concurrentSearchSettingsEnabled && requestShouldUseConcurrentSearch;
+    }
+
+    /**
+     * Evaluate if parsed request supports concurrent segment search
+     */
+    public void evaluateRequestShouldUseConcurrentSearch() {
```

Comment:
I believe this method is not needed: we could use `aggregations(...)` setter from the search context to decide if concurrent search could be used. This is the only place those are set.

Reply:
I'm having trouble finding the issue now but I believe @sohami is looking at disabling concurrent segment search for certain (non-agg) queries as well based on the request so I was coming at this from the perspective of introducing a way to determine if we should use concurrent segment search based on the request in general rather than just for aggs. @sohami please let me know if I'm wrong about this though and I'd be happy to make the change. Thanks!

Reply:
This is totally fine as well since we have `request` in the `DefaultSearchContext`  and could make this decision upfront, so there is no need to have dedicated method call to evaluate that

Reply:
@reta This method will be a single place to evaluate for request based conditions where we want to disable fallback to non concurrent path for the request. There are cases due to which using `request` passed in constructor doesn't work well. For example: 

1. In case of `scroll` request, the aggregations are not added in the `SearchContext` though it can be part of the `request` object used by `SearchContext`. [See the usage](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/SearchService.java#L1237) of `includeAggregations` flag in the `parseSource` method.
2. For case of timeseries index where sort request is made on timeseries field we also want to use non concurrent path. Details [here](https://github.com/opensearch-project/OpenSearch/issues/9538). To evaluate that we will need the `SortAndFormats` set in `SearchContext` after parsing the source. The condition will be same as used [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java#L527) in ContextIndexSearcher, which I will add as a follow-up to this change.

Reply:
> @reta This method will be a single place to evaluate for request based conditions where we want to fallback to non concurrent path for the request

Fair enough, should the at least accept the request it evaluates the conditions? For example: 

```
public void evaluateSearchConstraints(Request request) {
}
```

that would cleanly indicate two things a) request as the context (as you pointed out, the default is not enough) b) not that important, but could potentially evaluate any other restrictions. The concurrent search part should be represented as `SetOnce<Boolean>`

Reply:
We don't need `Request` since it is not used. The evaluation is done on the components stored in Context after parsing is complete. Like for `sort` evaluation will use `SortAndFormats` which is stored in context (after parsing) and not in `Request`.

> The concurrent search part should be represented as SetOnce<Boolean>

Makes sense

Reply:
Thanks @sohami 

> We don't need Request since it is not used. The evaluation is done on the components stored in Context after parsing is complete.

Do we have a deterministic place when this is done? (sorry haven't look yet). Why I am asking, do we have an opportunity to not have calling yet another method (at some place) but deduct from the context?

Reply:
So to summarize, @sohami  please correct me, by hooking up with:
 - `parseSource()` 
 - `aggregations(...)`  
 - `sort(...)`
 
we could decide (no manual method call) if the search request could use the concurrent search or not (right now we rely on `aggregations(...)` only but it will be extended later on). Is that right?

Reply:
No, I was still thinking to keep this method call `evaluateSearchConstraints()` in `parseSource` where it will evaluate in one place based on all request level parameters set in context. It will be more readable to keep this logic in single place instead of multiple methods like `aggregations` and `sort` and also follow the `SetOnce` semantic for the flag `requestShouldUseConcurrentSearch`. I will prefer to keep `sort` and `aggregations` methods simple setters and not have any other logic.

Reply:
> I will prefer to keep sort and aggregations methods simple setters and not have any other logic

:100: that would be ideal, but would aggregations + sort be set **before** parseSource?

Reply:
> but would aggregations + sort be set before parseSource?

No. Since these operators are part of request it gets set in `parseSource`

Reply:
Ok, so let's keep the logic this way but use `SetOnce`, thanks @sohami

Reply:
@reta @sohami thanks for the discussion here! I changed this to use `SetOnce` and changed the default case to set `true`.